File: Swift/Sources/buzz/scanner.swift
enum LexicalError {
    case unterminatedString(line: Int, column: Int)
    case unexpectedCharacter(line: Int, column: Int)
}

final class Scanner {
    let source: String
    
    var errors: [LexicalError] = []
    private var tokens: [Token] = []
    private var current: (start: Int, line: Int, column: Int, offset: Int) = (0, 0, 0, 0)
    
    init(source: String) {
        self.source = source
    }
    
    func scan() -> [Token] {
        while (!isEOF()) {
            current.start = current.offset
            scanToken()
        }
        
        tokens.append(Token(token: .eof, lexeme: "\0", literal: "\0" as Any, line: current.line, column: current.column))
        return tokens
    }
    
    private func scanToken() {
        let char = advance()
        switch (char) {
        case Character("|"):
            // Ignore comment
            while (peek() != Character("\n") && !isEOF()) {
                advance();
            }
        case Character("["):
            addToken(tokenType: .leftBracket)
        case Character("]"):
            addToken(tokenType: .rightBracket)
        case Character("("):
            addToken(tokenType: .leftParen)
        case Character(")"):
            addToken(tokenType: .rightParen)
        case Character("{"):
            addToken(tokenType: .leftBrace)
        case Character("}"):
            addToken(tokenType: .rightBrace)
        case Character(","):
            addToken(tokenType: .comma)
        case Character(";"):
            addToken(tokenType: .semicolon)
        case Character(">"):
            addToken(tokenType: match(Character("=")) ? .greaterEqual : .greater)
        case Character("<"):
            addToken(tokenType: match(Character("=")) ? .lessEqual : .less)
        case Character("+"):
            if match(Character("+")) {
                addToken(tokenType: .increment)
            } else if match(Character("=")) {
                addToken(tokenType: .plusEqual)
            } else {
                addToken(tokenType: .plus)
            }
        case Character("-"):
            if match(Character(">")) {
                addToken(tokenType: .arrow)
            } else if match(Character("-")) {
                addToken(tokenType: .decrement)
            } else if match(Character("=")) {
                addToken(tokenType: .minusEqual)
            } else {
                addToken(tokenType: .minus)
            }
        case Character("*"):
            if match(Character("=")) {
                addToken(tokenType: .starEqual)
            } else {
                addToken(tokenType: .star)
            }
        case Character("/"):
            addToken(tokenType: match(Character("=")) ? .slashEqual : .slash)
        case Character("%"):
            addToken(tokenType: .percent)
        case Character("?"):
            addToken(tokenType: match(Character("?")) ? .questionQuestion : .question)
        case Character("!"):
            addToken(tokenType: match(Character("=")) ? .bangEqual : .bang)
        case Character(":"):
            addToken(tokenType: .colon)
        case Character("="):
            addToken(tokenType: match(Character("=")) ? .equal : .equalEqual)
        case Character("\n"):
            current.line += 1
            current.column = 0
        case Character("\""):
            string()
        default:
            if char.isLetter {
                identifier()
            } else if char.isNumber {
                number()
            } else {
                errors.append(LexicalError.unexpectedCharacter(line: current.line, column: current.column))
            }
        }
    }
    
    private func identifier() {
        while (peek().isLetter) {
            advance()
        }
        
        let start = source.index(source.startIndex, offsetBy: current.start)
        let offset = source.index(source.startIndex, offsetBy: current.offset)
        let literal = String(source[start..<offset])
        
        for tokenType in TokenType.keywords {
            if tokenType.rawValue == literal {
                addToken(tokenType: tokenType)
                
                return
            }
        }
        
        addToken(tokenType: .identifier)
        return
    }
    
    private func number() {
        while (peek().isNumber) {
            advance()
        }
        
        if (peek() == Character(".") && peekNext().isNumber) {
            advance() // Consume .
            
            while (peek().isNumber) {
                advance()
            }
        }
        
        let start = source.index(source.startIndex, offsetBy: current.start)
        let offset = source.index(source.startIndex, offsetBy: current.offset)
        let literal = source[start..<offset]
        
        addToken(tokenType: .number, literal: Double(literal)! as Any)
    }
    
    private func string() {
        while (peek() != Character("\"") && !isEOF()) {
            if peek() == Character("\n") {
                errors.append(.unterminatedString(line: current.line, column: current.column))
            }
            
            advance()
        }
        
        if isEOF() {
            errors.append(.unterminatedString(line: current.line, column: current.column))
        } else {
            // Closing "
            advance()
        }
        
        // Trim quotes
        let start = source.index(source.startIndex, offsetBy: current.start + 1)
        let offset = source.index(source.startIndex, offsetBy: current.offset - 1)
        addToken(tokenType: .string, literal: String(source[start..<offset]) as Any)
    }
    
    private func isEOF() -> Bool {
        return current.offset >= source.count
    }
    
    private func peek() -> Character {
        if isEOF() {
            return Character("\0");
        }
        
        return charAt(current.offset)
    }
    
    private func peekNext() -> Character {
        if (current.offset + 1 > source.count) {
            return Character("\0")
        }
        
        return charAt(current.offset + 1)
    }
    
    @discardableResult
    private func advance() -> Character {
        let char = charAt(current.offset)
        current.offset += 1
        current.column += 1
        
        return char
    }
    
    private func match(_ expected: Character) -> Bool {
        if isEOF() {
            return false
        }
        
        if charAt(current.offset) != expected {
            return false
        }
        
        current.offset += 1
        return true
    }
    
    private func addToken(tokenType: TokenType, literal: Any? = nil) {
        let start = source.index(source.startIndex, offsetBy: current.start)
        let offset = source.index(source.startIndex, offsetBy: current.offset)
        
        let token = Token(
            token: tokenType,
            lexeme: String(source[start..<offset]),
            literal: literal,
            line: current.line,
            column: current.column
        )
        
        tokens.append(
            token
        )
        
        print("\(String(format: "%03d", token.line + 1)):\(String(format: "%03d", token.column + 1)) |\t\(token.token.rawValue.uppercased())\t\"\(token.lexeme)\"\(token.literal != nil ? "\t`\(token.literal!)`" : "")")
    }
    
    fileprivate func charAt(_ index: Int) -> Character {
        return source[source.index(source.startIndex, offsetBy: current.offset)]
    }
}

File: Rust/buzz/src/scanner.rs
use std::any::Any;

use crate::token::{Token, TokenType};
use unicode_segmentation::UnicodeSegmentation;

struct Location {
    start: usize,
    line: usize,
    column: usize,
    offset: usize,
}

pub struct Scanner<'a> {
    source: Vec<&'a str>,
    tokens: Vec<Token<'a>>,
    current: Location,
}

impl<'a> Scanner<'a> {
    fn scan(&'a mut self) {
        while !self.is_eof() {
            self.current.start = self.current.offset;
            self.scan_token();
        }

        self.tokens.push(Token {
            token: TokenType::Eof,
            lexeme: &self.source[self.source.len()..self.source.len() - 1],
            literal: Some(Box::new("\0")),
            location: (self.current.line, self.current.column),
        })
    }

    fn scan_token(&'a mut self) {
        let character = self.advance();
        match character {
            "|" => {
                while !self.peek().eq("\n") && !self.is_eof() {
                    self.advance();
                }
            }
            "[" => self.add_token(TokenType::LeftBracket, None),
            "]" => self.add_token(TokenType::RightBracket, None),
            "(" => self.add_token(TokenType::LeftParen, None),
            ")" => self.add_token(TokenType::LeftParen, None),
            "{" => self.add_token(TokenType::LeftBrace, None),
            "}" => self.add_token(TokenType::RightBrace, None),
            "," => self.add_token(TokenType::Comma, None),
            ";" => self.add_token(TokenType::Semicolon, None),
            ":" => self.add_token(TokenType::Colon, None),
            "%" => self.add_token(TokenType::Percent, None),
            ">" => {
                if self.match_char("=") {
                    self.add_token(TokenType::GreaterEqual, None);
                } else {
                    self.add_token(TokenType::Greater, None);
                }
            }
            "<" => {
                if self.match_char("=") {
                    self.add_token(TokenType::LessEqual, None);
                } else {
                    self.add_token(TokenType::Less, None);
                }
            }
            "+" => {
                if self.match_char("+") {
                    self.add_token(TokenType::Increment, None);
                } else if self.match_char("=") {
                    self.add_token(TokenType::PlusEqual, None);
                } else {
                    self.add_token(TokenType::Plus, None);
                }
            }
            "-" => {
                if self.match_char(">") {
                    self.add_token(TokenType::Arrow, None);
                } else if self.match_char("-") {
                    self.add_token(TokenType::Decrement, None);
                } else if self.match_char("=") {
                    self.add_token(TokenType::MinusEqual, None);
                } else {
                    self.add_token(TokenType::Minus, None);
                }
            }
            "*" => {
                if self.match_char("=") {
                    self.add_token(TokenType::StarEqual, None);
                } else {
                    self.add_token(TokenType::Star, None);
                }
            }
            "*" => {
                if self.match_char("=") {
                    self.add_token(TokenType::StarEqual, None);
                } else {
                    self.add_token(TokenType::Star, None);
                }
            }
            "/" => {
                if self.match_char("=") {
                    self.add_token(TokenType::SlashEqual, None);
                } else {
                    self.add_token(TokenType::Slash, None);
                }
            }
            "?" => {
                if self.match_char("?") {
                    self.add_token(TokenType::QuestionQuestion, None);
                } else {
                    self.add_token(TokenType::Question, None);
                }
            }
            "!" => {
                if self.match_char("=") {
                    self.add_token(TokenType::BangEqual, None);
                } else {
                    self.add_token(TokenType::Bang, None);
                }
            }
            "=" => {
                if self.match_char("=") {
                    self.add_token(TokenType::EqualEqual, None);
                } else {
                    self.add_token(TokenType::Equal, None);
                }
            }
            "\n" => {
                self.current.line += 1;
                self.current.column = 0;
            }
            "\"" => self.string(),
            _ => {
                eprintln!(
                    "Unexpected character at {}:{}",
                    self.current.line, self.current.column
                );
            }
        }
    }

    fn string(&'a mut self) {
        while !self.peek().eq("\"") && !self.is_eof() {
            if self.peek().eq("\n") {
                eprintln!(
                    "Unterminated string at {}:{}",
                    self.current.line, self.current.column
                );
            }

            self.advance();
        }

        if self.is_eof() {
            eprintln!(
                "Unterminated string at {}:{}",
                self.current.line, self.current.column
            );
        } else {
            // Closing "
            self.advance();
        }

        // self.add_token(
        //     TokenType::String,
        //     Some(Box::new(
        //         &self.source[(self.current.start + 1)..(self.current.offset - 1)],
        //     )),
        // )
    }

    fn is_eof(&self) -> bool {
        return self.current.offset > self.source.len();
    }

    fn peek(&self) -> &str {
        if self.is_eof() {
            return "\0";
        }

        return self.char_at(self.current.offset);
    }

    fn peek_next(&self) -> &str {
        if self.current.offset + 1 > self.source.len() {
            return "\0";
        }

        return self.char_at(self.current.offset);
    }

    fn advance(&mut self) -> &str {
        self.current.offset += 1;
        self.current.column += 1;

        return self.char_at(self.current.offset - 1);
    }

    fn char_at(&self, index: usize) -> &str {
        return self.source[index];
    }

    fn match_char(&mut self, expected: &str) -> bool {
        if self.is_eof() {
            return false;
        }

        if !self.char_at(self.current.offset).eq(expected) {
            return false;
        }

        self.current.offset += 1;
        return true;
    }

    fn add_token(&'a mut self, token_type: TokenType, literal: Option<Box<dyn Any>>) {
        let token = Token {
            token: token_type,
            lexeme: &self.source[self.current.start..(self.current.start + self.current.offset)],
            literal,
            location: (self.current.line, self.current.column),
        };

        self.tokens.push(token);
    }
}

pub fn scanner_from_string<'a>(source: &'a String) -> Scanner<'a> {
    return Scanner {
        source: source.graphemes(true).collect::<Vec<&str>>(),
        tokens: vec![],
        current: Location {
            start: 0,
            line: 0,
            column: 0,
            offset: 0,
        },
    };
}
